File: dxa.net/ecommerce-framework-dotnet/SDL.ECommerce.DXA/ECommerceRichTextProcessor.cs
using Sdl.Web.Common.Configuration;
using Sdl.Web.Common.Interfaces;
using Sdl.Web.Common.Logging;
using Sdl.Web.Common.Models;
using Sdl.Web.Tridion.Mapping;
using SDL.ECommerce.Api;
using SDL.ECommerce.DXA.Factories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.Web;

namespace SDL.ECommerce.DXA
{
    /// <summary>
    /// Rich text processor that are able to process product & category links.
    /// Is an extension of the OOTB rich text processor.
    /// To use this processor, please update your Unity.config.
    /// </summary>
    public class ECommerceRichTextProcessor : IRichTextProcessor
    {
        private static readonly Regex PRODUCT_URI_REGEX = new Regex("ecom:product:(?<productId>[^:]*):uri", RegexOptions.Compiled);
        private static readonly Regex CATEGORY_URI_REGEX = new Regex("ecom:category:(?<categoryId>[^:]*):uri", RegexOptions.Compiled);

        private readonly IECommerceLinkResolver _linkResolver;
        private readonly IRichTextProcessor _defaultRichTextProcessor;

        public ECommerceRichTextProcessor()
        {
            _linkResolver = DependencyFactory.Current.Resolve<IECommerceLinkResolver>();

            // TODO: Make it possible to inject the fallback rich text processor
            _defaultRichTextProcessor = new DefaultRichTextProcessor();
        }

        public RichText ProcessRichText(string xhtml, Localization localization)
        {
            xhtml = PRODUCT_URI_REGEX.Replace(xhtml, m => ResolveProductUrl(m.Groups["productId"].Value));
            xhtml = CATEGORY_URI_REGEX.Replace(xhtml, m => ResolveCategoryUrl(m.Groups["categoryId"].Value));
            return _defaultRichTextProcessor.ProcessRichText(xhtml, localization);
        }

        private string ResolveProductUrl(string productId)
        {
            var product = ECommerceContext.Client.DetailService.GetDetail(productId);
            return _linkResolver.GetProductDetailLink(product);
        }

        private string ResolveCategoryUrl(string categoryId)
        {
            var category = ECommerceContext.Client.CategoryService.GetCategoryById(categoryId);
            return _linkResolver.GetCategoryLink(category);
        }
        
    }
}

File: dxa.java/ecommerce-framework-dxa-module/src/main/java/com/sdl/ecommerce/dxa/ECommerceRichTextProcessor.java
package com.sdl.ecommerce.dxa;

import com.sdl.ecommerce.api.ECommerceLinkResolver;
import com.sdl.ecommerce.api.ProductCategoryService;
import com.sdl.ecommerce.api.ProductDetailResult;
import com.sdl.ecommerce.api.ProductDetailService;
import com.sdl.ecommerce.api.model.Category;
import com.sdl.webapp.common.api.localization.Localization;
import com.sdl.webapp.common.api.model.RichText;
import com.sdl.webapp.tridion.mapping.DefaultRichTextProcessor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Primary;
import org.springframework.stereotype.Component;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Primary
@Component
public class ECommerceRichTextProcessor extends DefaultRichTextProcessor {

    private static Pattern PRODUCT_URI_REGEX = Pattern.compile("ecom:product:([^:]+):uri");
    private static Pattern CATEGORY_URI_REGEX = Pattern.compile("ecom:category:([^:]+):uri");

    @Autowired
    private ProductCategoryService categoryService;

    @Autowired
    private ProductDetailService productDetailService;

    @Autowired
    private ECommerceLinkResolver linkResolver;

    @Override
    public RichText processRichText(String xhtml, Localization localization) {

        xhtml = processCategoryLinks(xhtml);
        xhtml = processProductLinks(xhtml);
        return super.processRichText(xhtml, localization);
    }

    private String processCategoryLinks(String xhtml) {
        Matcher matcher = CATEGORY_URI_REGEX.matcher(xhtml);
        StringBuffer sb = new StringBuffer();
        while (matcher.find()) {
            String categoryId = matcher.group(1);
            Category category = categoryService.getCategoryById(categoryId);
            if ( category != null ) {
                matcher.appendReplacement(sb, linkResolver.getCategoryLink(category));
            }
        }
        if ( sb.length() > 0 ) {
            matcher.appendTail(sb);
            xhtml = sb.toString();
        }
        return xhtml;
    }

    private String processProductLinks(String xhtml) {
        Matcher matcher = PRODUCT_URI_REGEX.matcher(xhtml);
        StringBuffer sb = new StringBuffer();
        while (matcher.find()) {
            String productId = matcher.group(1);
            ProductDetailResult result = productDetailService.getDetail(productId);
            if ( result != null && result.getProductDetail() != null ) {
                matcher.appendReplacement(sb, linkResolver.getProductDetailLink(result.getProductDetail()));
            }
        }
        if ( sb.length() > 0 ) {
            matcher.appendTail(sb);
            xhtml = sb.toString();
        }

        return xhtml;
    }
}
